ClassCastException when using BoxFolder.uploadLargeFile()

### Description of the Issue
When executing chunk upload by using BoxFolder.upload LargeFile(),
ClassCastException is raised often like the error stack as the following. 
This “often” means that sometimes the process completes successfully, sometimes I get this error.

### Cause of the bug
Please see [this file](https://cloud.box.com/s/x3tk9akir6f6cp0abnq7dzhvr0rwwjuz).

### Error Message, Including Stack Trace
Exception in thread "main" com.box.sdk.BoxAPIException: Unable to commit the upload session
at com.box.sdk.LargeFileUpload.uploadHelper(LargeFileUpload.java:204)
at com.box.sdk.LargeFileUpload.upload(LargeFileUpload.java:119)
at com.box.sdk.BoxFolder.uploadLargeFile(BoxFolder.java:1189)
at boxapp.UploadLargeTest2.main(UploadLargeTest2.java:28)
Caused by: java.lang.ClassCastException: class com.box.sdk.BoxAPIResponse cannot be cast to class com.box.sdk.BoxJSONResponse (com.box.sdk.BoxAPIResponse and com.box.sdk.BoxJSONResponse are in unnamed module of loader 'app')
at com.box.sdk.BoxJSONRequest.send(BoxJSONRequest.java:106)
at com.box.sdk.BoxJSONRequest.send(BoxJSONRequest.java:17)
at com.box.sdk.BoxAPIRequest.send(BoxAPIRequest.java:347)
at com.box.sdk.BoxJSONRequest.send(BoxJSONRequest.java:101)
at com.box.sdk.BoxFileUploadSession.commit(BoxFileUploadSession.java:192)
at com.box.sdk.LargeFileUpload.uploadHelper(LargeFileUpload.java:201)
... 3 more

### Versions Used
Java SDK:4.1.1

--- a/src/main/java/com/box/sdk/BoxAPIResponse.java
+++ b/src/main/java/com/box/sdk/BoxAPIResponse.java
@@ -5,7 +5,6 @@ import static com.box.sdk.http.ContentType.APPLICATION_JSON;
 import static java.lang.String.format;
 
 import com.eclipsesource.json.Json;
-import com.eclipsesource.json.JsonObject;
 import com.eclipsesource.json.ParseException;
 import java.io.ByteArrayInputStream;
 import java.io.Closeable;
@@ -44,7 +43,6 @@ public class BoxAPIResponse implements Closeable {
     private final String contentType;
     private final String requestMethod;
     private final String requestUrl;
-
     private int responseCode;
     private String bodyString;
 
@@ -150,6 +148,7 @@ public class BoxAPIResponse implements Closeable {
             );
         }
         ResponseBody responseBody = response.body();
+        //we must be better in identifying API vs JSON
         if (responseBody.contentLength() == 0 || responseBody.contentType() == null) {
             return new BoxAPIResponse(response.code(),
                 response.request().method(),
@@ -296,19 +295,6 @@ public class BoxAPIResponse implements Closeable {
         }
     }
 
-    public BoxJSONResponse toBoxJSONResponse() {
-        try {
-            return (BoxJSONResponse) this;
-        } catch (ClassCastException e) {
-            return new BoxJSONResponse(this.responseCode,
-                this.requestMethod,
-                this.requestUrl,
-                this.headers,
-                new JsonObject()
-            );
-        }
-    }
-
     /**
      * Returns a string representation of this response's body. This method is used when logging this response's body.
      * By default, it returns an empty string (to avoid accidentally logging binary data) unless the response contained
@@ -333,5 +319,13 @@ public class BoxAPIResponse implements Closeable {
         if (responseCode >= 500 && LOGGER.isErrorEnabled()) {
             LOGGER.error(this.toString());
         }
+    }
+
+    protected String getRequestMethod() {
+        return requestMethod;
+    }
+
+    protected String getRequestUrl() {
+        return requestUrl;
     }
 }

--- a/src/main/java/com/box/sdk/BoxAPIResponse.java
+++ b/src/main/java/com/box/sdk/BoxAPIResponse.java
@@ -298,7 +298,7 @@ public class BoxAPIResponse implements Closeable {
 
     public BoxJSONResponse toBoxJSONResponse() {
         try {
-            return (BoxJSONResponse)this;
+            return (BoxJSONResponse) this;
         } catch (ClassCastException e) {
             return new BoxJSONResponse(this.responseCode,
                 this.requestMethod,

--- a/src/main/java/com/box/sdk/BoxAPIResponse.java
+++ b/src/main/java/com/box/sdk/BoxAPIResponse.java
@@ -5,6 +5,7 @@ import static com.box.sdk.http.ContentType.APPLICATION_JSON;
 import static java.lang.String.format;
 
 import com.eclipsesource.json.Json;
+import com.eclipsesource.json.JsonObject;
 import com.eclipsesource.json.ParseException;
 import java.io.ByteArrayInputStream;
 import java.io.Closeable;

--- a/src/main/java/com/box/sdk/BoxAPIResponse.java
+++ b/src/main/java/com/box/sdk/BoxAPIResponse.java
@@ -295,6 +295,19 @@ public class BoxAPIResponse implements Closeable {
         }
     }
 
+    public BoxJSONResponse toBoxJSONResponse() {
+        try {
+            return (BoxJSONResponse)this;
+        } catch (ClassCastException e) {
+            return new BoxJSONResponse(this.responseCode,
+                this.requestMethod,
+                this.requestUrl,
+                this.headers,
+                new JsonObject()
+            );
+        }
+    }
+
     /**
      * Returns a string representation of this response's body. This method is used when logging this response's body.
      * By default, it returns an empty string (to avoid accidentally logging binary data) unless the response contained

--- a/src/main/java/com/box/sdk/BoxAPIResponse.java
+++ b/src/main/java/com/box/sdk/BoxAPIResponse.java
@@ -148,7 +148,6 @@ public class BoxAPIResponse implements Closeable {
             );
         }
         ResponseBody responseBody = response.body();
-        //we must be better in identifying API vs JSON
         if (responseBody.contentLength() == 0 || responseBody.contentType() == null) {
             return new BoxAPIResponse(response.code(),
                 response.request().method(),

--- a/src/main/java/com/box/sdk/BoxJSONRequest.java
+++ b/src/main/java/com/box/sdk/BoxJSONRequest.java
@@ -98,12 +98,12 @@ public class BoxJSONRequest extends BoxAPIRequest {
 
     @Override
     public BoxJSONResponse send() {
-        return (BoxJSONResponse) super.send();
+        return super.send().toBoxJSONResponse();
     }
 
     @Override
     public BoxJSONResponse send(ProgressListener listener) {
-        return (BoxJSONResponse) super.send(listener);
+        return super.send(listener).toBoxJSONResponse();
     }
 
     @Override

--- a/src/main/java/com/box/sdk/BoxJSONRequest.java
+++ b/src/main/java/com/box/sdk/BoxJSONRequest.java
@@ -98,12 +98,20 @@ public class BoxJSONRequest extends BoxAPIRequest {
 
     @Override
     public BoxJSONResponse send() {
-        return super.send().toBoxJSONResponse();
+        return convert(super.send());
     }
 
     @Override
     public BoxJSONResponse send(ProgressListener listener) {
-        return super.send(listener).toBoxJSONResponse();
+        return convert(super.send(listener));
+    }
+
+    private BoxJSONResponse convert(BoxAPIResponse response) {
+        if (response instanceof BoxJSONResponse) {
+            return (BoxJSONResponse) response;
+        } else {
+            return new BoxJSONResponse(response);
+        }
     }
 
     @Override

--- a/src/main/java/com/box/sdk/BoxJSONResponse.java
+++ b/src/main/java/com/box/sdk/BoxJSONResponse.java
@@ -25,6 +25,16 @@ public class BoxJSONResponse extends BoxAPIResponse {
         super();
     }
 
+    BoxJSONResponse(BoxAPIResponse response) {
+        this(
+            response.getResponseCode(),
+            response.getRequestMethod(),
+            response.getRequestUrl(),
+            response.getHeaders(),
+            new JsonObject()
+        );
+    }
+
     /**
      * Constructs a BoxAPIResponse with an http response code and response body.
      *

--- a/src/test/java/com/box/sdk/BoxJSONRequestTest.java
+++ b/src/test/java/com/box/sdk/BoxJSONRequestTest.java
@@ -1,0 +1,45 @@
+package com.box.sdk;
+
+import static com.box.sdk.TestUtils.createConnectionWith;
+import static com.github.tomakehurst.wiremock.client.WireMock.aResponse;
+import static com.github.tomakehurst.wiremock.client.WireMock.get;
+import static com.github.tomakehurst.wiremock.client.WireMock.stubFor;
+import static com.github.tomakehurst.wiremock.client.WireMock.urlEqualTo;
+import static com.github.tomakehurst.wiremock.core.WireMockConfiguration.wireMockConfig;
+import static java.lang.String.format;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.Is.is;
+import static org.mockito.Mockito.mock;
+
+import com.github.tomakehurst.wiremock.junit.WireMockRule;
+import java.net.MalformedURLException;
+import java.net.URL;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class BoxJSONRequestTest {
+    @Rule
+    public WireMockRule wireMockRule = new WireMockRule(wireMockConfig().dynamicHttpsPort().httpDisabled(true));
+
+    @Test
+    public void shouldHandleApiResponse() {
+        stubFor(get(urlEqualTo("/")).willReturn(aResponse().withStatus(202)));
+        Time mockTime = mock(Time.class);
+        BackoffCounter backoffCounter = new BackoffCounter(mockTime);
+
+        BoxAPIConnection api = createConnectionWith(boxMockUrl().toString());
+        BoxJSONRequest request = new BoxJSONRequest(api, boxMockUrl(), "GET");
+        request.setBackoffCounter(backoffCounter);
+
+        BoxJSONResponse response = request.send();
+        assertThat(response.getResponseCode(), is(202));
+    }
+
+    private URL boxMockUrl() {
+        try {
+            return new URL(format("https://localhost:%d/", wireMockRule.httpsPort()));
+        } catch (MalformedURLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+}

